File: frontend/src/api/filaments.ts
import api from '@/lib/api';
import type { Filament, FilamentType, FilamentColor } from '@/types';

export const filamentsApi = {
  list: () => api.get<Filament[]>('/api/filaments').then(res => res.data),
  create: (data: Partial<Filament>) => api.post<Filament>('/api/filaments', data).then(res => res.data),
  update: (id: number, data: Partial<Filament>) => api.patch<Filament>(`/api/filaments/${id}`, data).then(res => res.data),
  delete: (id: number) => api.delete(`/api/filaments/${id}`).then(res => res.data),
  
  listTypes: () => api.get<FilamentType[]>('/api/filaments/types').then(res => res.data),
  createType: (name: string) => api.post<FilamentType>('/api/filaments/types', { name, is_custom: true }).then(res => res.data),
  deleteType: (id: number) => api.delete(`/api/filaments/types/${id}`).then(res => res.data),
  
  listColors: () => api.get<FilamentColor[]>('/api/filaments/colors').then(res => res.data),
  createColor: (data: Partial<FilamentColor>) => api.post<FilamentColor>('/api/filaments/colors', data).then(res => res.data),
  deleteColor: (id: number) => api.delete(`/api/filaments/colors/${id}`).then(res => res.data),
};

File: backend/app/routers/filaments.py
from typing import List
from fastapi import APIRouter, Depends
from sqlmodel import Session
from app.dependencies import get_session
from app.services.filament_service import FilamentService
from app.schemas.filament import (
    FilamentRead,
    FilamentCreate,
    FilamentUpdate,
    FilamentTypeRead,
    FilamentTypeCreate,
    FilamentColorRead,
    FilamentColorCreate,
)

router = APIRouter(prefix="/filaments", tags=["filaments"])


@router.get("", response_model=List[FilamentRead])
def list_filaments(session: Session = Depends(get_session)):
    return FilamentService.list_filaments(session)


@router.post("", response_model=FilamentRead)
def create_filament(data: FilamentCreate, session: Session = Depends(get_session)):
    return FilamentService.create_filament(session, data)


@router.get("/{filament_id}", response_model=FilamentRead)
def get_filament(filament_id: int, session: Session = Depends(get_session)):
    return FilamentService.get_filament(session, filament_id)


@router.patch("/{filament_id}", response_model=FilamentRead)
def update_filament(
    filament_id: int, data: FilamentUpdate, session: Session = Depends(get_session)
):
    return FilamentService.update_filament(session, filament_id, data)


@router.delete("/{filament_id}")
def delete_filament(filament_id: int, session: Session = Depends(get_session)):
    FilamentService.delete_filament(session, filament_id)
    return {"ok": True}


# Types
@router.get("/types", response_model=List[FilamentTypeRead])
def list_types(session: Session = Depends(get_session)):
    return FilamentService.list_types(session)


@router.post("/types", response_model=FilamentTypeRead)
def create_type(data: FilamentTypeCreate, session: Session = Depends(get_session)):
    return FilamentService.create_type(session, data)


@router.delete("/types/{type_id}")
def delete_type(type_id: int, session: Session = Depends(get_session)):
    FilamentService.delete_type(session, type_id)
    return {"ok": True}


# Colors
@router.get("/colors", response_model=List[FilamentColorRead])
def list_colors(session: Session = Depends(get_session)):
    return FilamentService.list_colors(session)


@router.post("/colors", response_model=FilamentColorRead)
def create_color(data: FilamentColorCreate, session: Session = Depends(get_session)):
    return FilamentService.create_color(session, data)


@router.delete("/colors/{color_id}")
def delete_color(color_id: int, session: Session = Depends(get_session)):
    FilamentService.delete_color(session, color_id)
    return {"ok": True}
